pkg/gatewayserver/io/ws/ws_util_test.go

```diff
@@ -38,6 +39,38 @@ import (
 	"go.thethings.network/lorawan-stack/v3/pkg/util/test"
 )
 
+// mockClock is a mock gateway clock.
+// It rolls over every 32 bits, approximately every 72 minutes.
+type mockClock struct {
+	sessionID      uint16
+	startTimeStamp uint32
+	startTime      time.Time
+}
+
+// Start starts the gateway clock.
+func (c *mockClock) Start(ctx context.Context, now time.Time) {
+	// Set the clock to a random value on every start.
+	c.startTimeStamp = rand.Uint32()
+	c.sessionID = uint16(rand.Uint32())
+	c.startTime = now
+}
+
+// GetTimestamp returns the current timestamp.
+func (c *mockClock) GetTimestamp() uint32 {
+	return uint32(int64(c.startTimeStamp) + time.Since(c.startTime).Microseconds())
+}
+
+// GetXTime returns the timestamp in the LoRa Basics Station `xtime` format for the given time.
+func (c *mockClock) GetXTimeForTime(t time.Time) int64 {
+	ts := uint32(int64(c.startTimeStamp) + t.Sub(c.startTime).Microseconds())
+	return int64(c.sessionID)<<48 | int64(ts)&0xFFFFFFFFFFFF
```

Comment:
Maybe the helpers for time conversion should be moved such that they are accessible here. We're copy pasting too much these formulas and if we're some `FF` off it is bad.

Reply:
Yeah so I can't just copy the helpers. This mock uses a microsecond timestamp that simulates a gateway and `scheduling.ConcentratorTime` is in nanoseconds. I can move it to the helper package but this is too specific to the test code that I prefer to leave this here.

Reply:
You can move the constant (`0xFFFFFFFFFFFF`).